src/Tools/ExternalAccess/LiveShare/Classification/RoslynSyntaxClassificationService.cs

```diff
@@ -0,0 +1,192 @@
+    {
+        private readonly AbstractLspClientServiceFactory _roslynLspClientServiceFactory;
+        private readonly RemoteLanguageServiceWorkspace _remoteLanguageServiceWorkspace;
+        private readonly ISyntaxClassificationService _originalService;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IExperimentationService _experimentationService;
+        private readonly IThreadingContext _threadingContext;
+
+        public RoslynSyntaxClassificationService(AbstractLspClientServiceFactory roslynLspClientServiceFactory, RemoteLanguageServiceWorkspace remoteLanguageServiceWorkspace, ISyntaxClassificationService originalService,
+            ClassificationTypeMap classificationTypeMap, IExperimentationService experimentationService, IThreadingContext threadingContext)
+        {
+            _roslynLspClientServiceFactory = roslynLspClientServiceFactory;
+            _remoteLanguageServiceWorkspace = remoteLanguageServiceWorkspace;
+            _originalService = originalService;
+            _classificationTypeMap = classificationTypeMap;
+            _experimentationService = experimentationService;
+            _threadingContext = threadingContext;
+        }
+
+        public void AddLexicalClassifications(SourceText text, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Local))
+            {
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+                }
+            }
+            else if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Remote))
+            {
+                var documentId = _remoteLanguageServiceWorkspace.GetDocumentIdInCurrentContext(text.Container);
+                var document = _remoteLanguageServiceWorkspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    // It is expected that the document cannot be found (e.g. source text comes from preview workspace).
+                    // These cases are already not supported in liveshare, so for now just return.
+                    return;
+                }
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _threadingContext.JoinableTaskFactory.Run(async () =>
+                    {
+                        await AddRemoteClassificationsAsync(LexicalClassificationsHandler.LexicalClassificationsMethodName, document.FilePath, text, textSpan, result, cancellationToken).ConfigureAwait(false);
+                    });
```

Comment:
it's very very very intentional that lexical classificatio is synchronous.  even small experiments with doing this async proved very unpleasant for people.  i.e. you didn't get the "snap" of keywords turning blue (or not turning blue if it became a non-keyword).  etc. etc.

Note that OOP experiments for async syntactic and semantic classification seemed ok for people.

Reply:
does anythnig boost the priority of this to *max* in whatever communication system we use?

Reply:
this is live share case where connection is cross internet and we are trying to see supporting live share without people ever installing Roslyn workload in client side. we will see whether people care to install Roslyn workload just to have liveshare with faster "keyword" highlighting or they are okay with slower keyword highlighting as long as they can join liveshare as fast/cheaply as they can.